### PR TITLE
neonvm-runner: arm support

### PR DESCRIPTION
**Pull request by mikhail-sakhnov**
Modify qemu arguments to run VM on arm machine

#1080

### CODE REVIEW

File: neonvm-runner/cmd/main.go [hunk 1]
```diff
@@ -625,9 +631,10 @@ func newConfig(logger *zap.Logger) *Config {
 		appendKernelCmdline:  "",
 		skipCgroupManagement: false,
 		diskCacheSettings:    "cache=none",
-		memoryProvider:       "", // Require that this is explicitly set. We'll check later.
-		autoMovableRatio:     "", // Require that this is explicitly set IFF memoryProvider is VirtioMem. We'll check later.
-		cpuScalingMode:       "", // Require that this is explicitly set. We'll check later.
+		memoryProvider:       "",             // Require that this is explicitly set. We'll check later.
+		autoMovableRatio:     "",             // Require that this is explicitly set IFF memoryProvider is VirtioMem. We'll check later.
+		cpuScalingMode:       "",             // Require that this is explicitly set. We'll check later.
+		architecture:         runtime.GOARCH, // arm64, amd64
```
Comment (sharnoff) [inline neonvm-runner/cmd/main.go:637]: super tiny nit: can we put this elsewhere in the struct so it doesn't expand the diff?

File: neonvm-runner/cmd/main.go [hunk 2]
```diff
@@ -1104,6 +1111,28 @@ func runQEMU(
...
+	case architectureArm64:
+		return qemuBinArm64
+	case architectureAmd64:
+		return qemuBinX8664
+	default:
+		panic(fmt.Errorf("unknown architecture %s", architecture))
+	}
+}
+
+func getMachineType(architecture string) string {
```
Comment (sharnoff) [inline neonvm-runner/cmd/main.go:1125]: It'd be good to have a comment explaining the situation here -- why these particular machine types?

File: tests/e2e/autoscaling.cpu-sys-fs-state-scaling/00-create-vm.yaml [hunk 3]
```diff
@@ -37,7 +37,7 @@ spec:
       min: 1
       max: 5
       use: 1
-    memoryProvider: DIMMSlots
+    memoryProvider: VirtioMem
```
Comment (sharnoff) [inline tests/e2e/autoscaling.cpu-sys-fs-state-scaling/00-create-vm.yaml:40]: Should this change be merged into the PR this is based on?

(+ same question for the other changes to this test)

**Review by sharnoff — COMMENTED** [review]
Looks good, AFAICT

File: neonvm-runner/cmd/main.go (hunk 1 above)
Comment (mikhail-sakhnov) [reply neonvm-runner/cmd/main.go:637]: I put it here as all other fields have for the sake of similarity.  Happy to move all three comments to the structure definition though.

File: neonvm-runner/cmd/main.go (hunk 2 above)
Comment (mikhail-sakhnov) [reply neonvm-runner/cmd/main.go:1125]: tbh there are no particular reason. for q35 it is probably historical reason and virt is the most generic arm platform which we use in runners already.

Comment (sharnoff) [reply neonvm-runner/cmd/main.go:1129]: perhaps @cicdteam can add some context on why q35?

File: tests/e2e/autoscaling.cpu-sys-fs-state-scaling/00-create-vm.yaml (hunk 3 above)
Comment (sharnoff) [reply tests/e2e/autoscaling.cpu-sys-fs-state-scaling/00-create-vm.yaml:40]: Fair enough, makes sense

File: neonvm-runner/cmd/main.go (hunk 2 above)
Comment (cicdteam) [reply neonvm-runner/cmd/main.go:1129]: Sorry, I just saw this discussion.

In the x86_64 architecture, QEMU has two groups of emulated machines - `pc-q35` and `pc-i440fx` (yes, there are a couple of others, like `microvm` or `none`, but this is not our case). The `q35` series is more recent, so it was chosen. In fact, there are no other options for x86_64. 

FYI: `-machine q35` is alias of `-machine pc-q35-9.0` (in QEMU v9.0.1),by other words - most recent hardware.

Comment (mikhail-sakhnov) [reply neonvm-runner/cmd/main.go:1150]: added some clarifications on that

File: pkg/neonvm/controllers/vm_controller.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ import (
 	"fmt"
 	"os"
 	"reflect"
+	sysruntime "runtime"
```
Comment (Omrigan) [inline pkg/neonvm/controllers/vm_controller.go:29]: Why non-standard name?

File: pkg/neonvm/controllers/vm_controller.go [hunk 5]
```diff
@@ -1222,7 +1222,7 @@ func affinityForVirtualMachine(vm *vmv1.VirtualMachine) *corev1.Affinity {
 					{
 						Key:      "kubernetes.io/arch",
 						Operator: "In",
-						Values:   []string{"amd64"},
```
Comment (Omrigan) [inline pkg/neonvm/controllers/vm_controller.go:1225]: What's the reasoning behind this change,. why does VM's affinity must match controller's architecture?

If we might have heterogeneous clusters, maybe we need to specify the architecture explicitly in the spec?

File: neonvm-vxlan-controller/Dockerfile
Comment (Omrigan) [inline neonvm-vxlan-controller/Dockerfile:1]: vxlan-controller changes should be a separate commit.

**Review by Omrigan — COMMENTED** [review]
Looking good!

Perhaps, PR should be named something like "enable e2e tests on arm", with multiple commits, 1 per each component?

File: pkg/neonvm/controllers/vm_controller.go (hunk 5 above)
Comment (mikhail-sakhnov) [reply pkg/neonvm/controllers/vm_controller.go:1225]: We already can specify it through the VM spec, by passing affinity. 

I think at the moment using by default the same arch as the controller might be a safe assumption. Do you have any other options in your mind?

File: pkg/neonvm/controllers/vm_controller.go (hunk 4 above)
Comment (mikhail-sakhnov) [reply pkg/neonvm/controllers/vm_controller.go:29]: The same file also has `"k8s.io/apimachinery/pkg/runtime"` import

**Comment by mikhail-sakhnov** [comment]
> Perhaps, PR should be named something like "enable e2e tests on arm", with multiple commits, 1 per each component?

@Omrigan I am hesitant towards that title, because the PR does what it says - enables vm-runner to run ARM vms. There is one test modified to reflect that change. Plus there actually separate issue and PR about e2e tests on ARM.

File: neonvm-vxlan-controller/Dockerfile
Comment (mikhail-sakhnov) [reply neonvm-vxlan-controller/Dockerfile:1]: I have feeling that it is a separate commit at the moment https://github.com/neondatabase/autoscaling/commit/59ccc8965da891958f09606a5f2fe6888f351b23

Comment (Omrigan) [reply neonvm-vxlan-controller/Dockerfile:1]: Agreed to extract it: https://github.com/neondatabase/autoscaling/pull/1160.

File: pkg/neonvm/controllers/vm_controller.go (hunk 5 above)
Comment (Omrigan) [reply pkg/neonvm/controllers/vm_controller.go:1225]: > Maybe it worth to discuss whether we want to have heterogeneous clusters.

The existence of https://github.com/neondatabase/autoscaling/pull/1123 hints toward yes. In that case, we need to align on how it should be determined where to schedule the VM.

Comment (mikhail-sakhnov) [reply pkg/neonvm/controllers/vm_controller.go:1226]: @Omrigan do you have any suggestions?

**Comment by mikhail-sakhnov** [comment]
@Omrigan @sharnoff what is the current state of review?

File: pkg/neonvm/controllers/vm_controller.go (hunk 5 above)
Comment (mikhail-sakhnov) [reply pkg/neonvm/controllers/vm_controller.go:1226]: we agreed verbally with @sharnoff to keep it as current PR has, but change PR #1083 with explicit spec.arch field

**Review by Omrigan — APPROVED** [review]

File: neonvm-runner/Dockerfile [hunk 6]
```diff
@@ -21,12 +20,16 @@ RUN apk add --no-cache \
     busybox-extras \
     e2fsprogs \
     qemu-system-x86_64 \
+    qemu-system-aarch64 \
```
Comment (sharnoff) [inline neonvm-runner/Dockerfile:23]: question: What is the expected behavior when installing either of these on the other architecture?

(is it worth only installing one, depending on target arch?)

File: neonvm-runner/cmd/main.go [hunk 7]
```diff
@@ -851,6 +861,19 @@ func buildQEMUCmd(
...
+		qemuCmd = append(qemuCmd, "-bios", "/vm/QEMU_EFI_ARM.fd")
+		// arm virt has only one UART, setup virtio-serial to add more /dev/hvcX
+		qemuCmd = append(qemuCmd,
+			"-chardev", "stdio,id=virtio-console",
+			"-device", "virtconsole,chardev=virtio-console",
+		)
+	} else {
+		// on amd we have multiple UART ports so we can just use serial stdio
+		qemuCmd = append(qemuCmd, "-serial", "stdio")
+	}
```
Comment (sharnoff) [inline neonvm-runner/cmd/main.go:875]: suggestion: let's use `switch` for these and panic in the default case, so it's more clear which architecture gets which behavior?

(+ same suggestion for other occurrences of this)

File: neonvm-runner/cmd/main.go [hunk 8]
```diff
@@ -625,9 +635,10 @@ func newConfig(logger *zap.Logger) *Config {
 		appendKernelCmdline:  "",
 		skipCgroupManagement: false,
 		diskCacheSettings:    "cache=none",
-		memoryProvider:       "", // Require that this is explicitly set. We'll check later.
-		autoMovableRatio:     "", // Require that this is explicitly set IFF memoryProvider is VirtioMem. We'll check later.
-		cpuScalingMode:       "", // Require that this is explicitly set. We'll check later.
+		memoryProvider:       "",
+		autoMovableRatio:     "",
+		cpuScalingMode:       "",
+		architecture:         runtime.GOARCH,
```
Comment (sharnoff) [inline neonvm-runner/cmd/main.go:641]: question: Is this only ever equal to `runtime.GOARCH` ? Do we plan to keep it that way? (if so: can we make assumptions that e.g. we're only ever executing `qemu-system-x86_64` from `amd64` go code?)

**Review by sharnoff — COMMENTED** [review]
Left a couple comments. I may have some follow-up questions pending the outcome of the `runtime.GOARCH` one, but should be nothing major -- broadly looks good

File: neonvm-runner/cmd/main.go (hunk 8 above)
Comment (mikhail-sakhnov) [reply neonvm-runner/cmd/main.go:641]: Yes, and I think I got where are you heading to - why not to use `runtime.GOARCH` everywhere?

Comment (mikhail-sakhnov) [reply neonvm-runner/cmd/main.go:647]: >But also — why install both qemu-system-x86_64 and qemu-system-aarch64 if we're only going to need one? (and then if we like, we can set up some symlink to refer to it, so we don't need to match the name inside the go code)



Basically to simplify the image builds, since this PR was introduced way before crossplatform images. I can update PR with crossplatform images to use only the correct version of the qemu.

But in general, cfg.architecture is used in a couple of places (kernel arguments, qemu arguments (not binary name), adding firmware) and I don't want to make it using `runtime.GOARCH` everywhere instead. 

so: 
- I will update the PR with #1138 to have only one qemu binary
- I keep cfg.architecture here, for the sake of using it in different places

How does it sound?

File: neonvm-runner/Dockerfile (hunk 6 above)
Comment (mikhail-sakhnov) [reply neonvm-runner/Dockerfile:23]: As mentioned in another thread, it was done this way to simplify builds. There is a PR to make proper cross-platform images, which might be extended with changes to install only platform-specific qemu package.

File: neonvm-runner/cmd/main.go (hunk 7 above)
Comment (mikhail-sakhnov) [reply neonvm-runner/cmd/main.go:966]: changed to switch

File: neonvm-runner/cmd/main.go (hunk 8 above)
Comment (sharnoff) [reply neonvm-runner/cmd/main.go:647]: Fair enough, sounds like a plan :)

**Review by sharnoff — APPROVED** [review]
